books/compute/src/cuda/kernels/layernorm_forward.md

```diff
@@ -1,11 +1,494 @@
+    const float* x = inp + idx * C;
+    // thread coarsening
+    float sum = 0.0f;
+    for (int i = tid; i < C; i += block_size) {
+        sum += x[i];
+    }
+    shared[tid] = sum;
+    __syncthreads();
+    // reductions
+    for (int stride = block_size / 2; stride >= 1; stride /= 2) {
+        __syncthreads();
+        if (tid < stride) {
+            shared[tid] += shared[tid + stride];
+        }
+    }
+    // write the final result (at thread 0) to global memory
+    if (tid == 0) {
+        mean[idx] = shared[0] / C;
+    }
+}
+
+__global__ void rstd_kernel(
+    float* rstd,
+    const float* inp, const float* mean,
+    int N, int C,
+    int block_size
+) {
+    extern __shared__ float shared[];
+    int idx = blockIdx.x; // range [0, B*T)
+    int tid = threadIdx.x; // range [0, block_size)
+    const float* x = inp + idx * C;
+    float m = mean[idx];
+    // thread coarsening
+    float sum = 0.0f;
+    for (int i = tid; i < C; i += block_size) {
+        float diff = x[i] - m;
+        sum += diff * diff;
+    }
+    shared[tid] = sum;
+    __syncthreads();
+    // reductions
+    for (int stride = block_size / 2; stride >= 1; stride /= 2) {
+        __syncthreads();
+        if (tid < stride) {
+            shared[tid] += shared[tid + stride];
+        }
+    }
+    // write the final result (at thread 0) to global memory
+    if (tid == 0) {
+        rstd[idx] = 1.0f / sqrtf(shared[0] / C + 1e-5f);
+    }
+}
+
+__global__ void normalization_kernel(
+    float* out,
+    const float* inp, float* mean, float* rstd,
+    const float* weight, const float* bias,
+    int B, int T, int C
+) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    int bt = idx / C;
+    int c = idx % C;
+
+    float m = mean[bt];
+    float s = rstd[bt];
+    float xi = inp[idx];
+    float n = s * (xi - m);
+    float o = n * weight[c] + bias[c];
+
+    out[idx] = o;
+}
+
+void layernorm_forward2(
+    float* out, float* mean, float* rstd,
+    const float* inp, const float* weight, const float* bias,
+    int B, int T, int C,
+    const int block_size
+) {
+    int N = B * T;
+
+    // in mean and rstd, threads cooperate within blocks via reductions
+    mean_kernel<<<N, block_size, block_size * sizeof(float)>>>(
+        mean, inp, N, C, block_size
+    );
+    cudaCheck(cudaGetLastError());
+    rstd_kernel<<<N, block_size, block_size * sizeof(float)>>>(
+        rstd, inp, mean, N, C, block_size
+    );
+    cudaCheck(cudaGetLastError());
+
+    // in the normalization, everything just gets flattened out
+    const int block_size2 = 256;
+    const int grid_size = ceil_div(B * T * C, block_size2);
+    normalization_kernel<<<grid_size, block_size2>>>(
+        out, inp, mean, rstd, weight, bias, B, T, C
+    );
+    cudaCheck(cudaGetLastError());
+}
+```
+
+## Kernel 3
+
+Kernel 3 introduces the use of *cooperative groups*, which allows us to utilize
+thread groups of arbitrary sizes (multiples of 2) and not limited to the thread
```

Comment:
Readability edit:
- "...  rather than being restricted to the thread block size."